jupyterhub/singleuser/mixins.py

```diff
@@ -707,6 +715,26 @@ def make_singleuser_app(App):
                 "{}.base_handler_class must be defined".format(App.__name__)
             )
 
+    # patch-in hub HubOAuthCallbackHandler to BaseHandler,
+    # so anything inheriting from BaseHandler uses Hub authentication
+    if HubAuthenticatedHandler not in BaseHandler.__bases__:
+        log.debug(f"Patching {HubAuthenticatedHandler} into {BaseHandler}")
+        BaseHandler.__bases__ = (HubAuthenticatedHandler,) + BaseHandler.__bases__
+        # adding it to bases isn't enough to override methods defined on BaseHandler, though.
+        # we still need to override any methods defined on BaseHandler *itself* that we should override
+        # since bases come immediately *after* the class itself
+        # as of writing, there are no collisions on the default classes
+        # so this block has no effect for ServerApp or NotebookApp
```

Comment:
Is it correct that JupyterLab 3 will rely on another App defined somewhere outside the jupyterhub package? I thought it was going to rely on ServerApp defined by jupyter_server instead of NotebookApp, but if `this block has no effect...` then JupyterLab 3 will still have the same issues?

Oh, wait, what is the `block`? Can you clarify the comment so it is more apparent what `block` refers to? I consider if its the entire if statement or what is below the comment only.

Reply:
JupyterLab does use ServerApp, yes.

Block means a section of code. It refers to the block of code handling methods defined on BaseHandler.

Patching into `__bases__` effectively rewrites the definition:

```python
class BaseHandler(Parent1, Parent1):
    def foo(self):
    ....
```

to

```python
class BaseHandler(HubAuthenticatedHandler, Parent1, Parent2):
    def foo(self):
    ....
```

as we expect from inheritance, this means methods defined in HubAuthenticatedHandler override those defined in the original Parents, but not those defined in BaseHandler itself. In the above example, `foo` would not be overridden, but we need it to be if it is one of the methods or attributes used by HubAuthenticatedHandler. That's what this extra block handles.

I'll have a go at simplifying it to be more explicit about the two public methods that *might* be defined here, and ignore the other methods and attributes in the mixin that are hub-specific.

This block is out of an abundance of caution/correctness, because I added this comment:

> this block has no effect for ServerApp or NotebookApp

when I discovered this situation does not occur with the packages as they are now. None of these methods are defined on BaseHandler, they are all inherited, so inserting into `__bases__` *happens* to be enough. If, however, the upstream package moved the definition of `get_current_user` to `BaseHandler` itself, without this block, auth would not be overridden at all, which is what I want to avoid.

I've updated the code and comment to be a bit more direct and narrow, and hopefully clearer.

Reply:
update: actually pushed it just now

Reply:
> ```python
> class BaseHandler(HubAuthenticatedHandler, Parent1, Parent2):
>     def foo(self):
>     ....
> ```
> 
> as we expect from inheritance, this means methods defined in HubAuthenticatedHandler override those defined in the original Parents, but not those defined in BaseHandler itself. In the above example, `foo` would not be overridden, but we need it to be if it is one of the methods or attributes used by HubAuthenticatedHandler. That's what this extra block handles.

This made me "ahaaa!"!

---

I don't feel I understand all the pieces in play still though because I'm not very knowledable about ServerApp and NotebookApp yet. With all this in mind this LGTM given what I understand and suggest you merge this if you feel its ready to go!

Thank you for helping me learn about these parts @minrk!